File: XCTestBootstrap/Utility/FBOToolDynamicLibs.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBOToolDynamicLibs.h"

#import "FBOToolOperation.h"

@implementation FBOToolDynamicLibs

+ (FBFuture<NSArray *> *)findFullPathForSanitiserDyldInBundle:(NSString *)bundlePath onQueue:(nonnull dispatch_queue_t)queue {
    return [[FBOToolOperation listSanitiserDylibsRequiredByBundle:bundlePath onQueue:queue] onQueue:queue map:^id _Nonnull(NSArray<NSString *> * _Nonnull libsList) {

        NSString *clanLocation = [FBXcodeConfiguration.developerDirectory stringByAppendingPathComponent:@"Toolchains/XcodeDefault.xctoolchain/usr/lib/clang"];
        NSError *error = nil;
        NSArray<NSURL *> *fileList = [self filesInDirectory:[NSURL fileURLWithPath:clanLocation] error:&error];

        if ([fileList count] == 0 || error) {
            if(error == nil) return [[FBControlCoreError
                                      describeFormat:@"No clang version found in %@", clanLocation] failFuture];
            return [FBFuture futureWithError:error];
        }

        NSString *libsFolder = [NSString pathWithComponents: @[[fileList[0] path], @"lib/darwin/"]];

        NSString *bundleFrameworksFolder = [bundlePath stringByAppendingPathComponent:@"Frameworks"];
        if(![[NSFileManager defaultManager] fileExistsAtPath:bundleFrameworksFolder]) {
          bundleFrameworksFolder = [bundlePath stringByAppendingPathComponent:@"Contents/Frameworks"];
        }

        NSArray<NSURL *> *bundleLibs = [self filesInDirectory:[NSURL fileURLWithPath:bundleFrameworksFolder] error:&error];
        NSMutableSet<NSString *> *bundleLibsNames = nil;
        if (bundleLibs) {
          bundleLibsNames = [NSMutableSet setWithCapacity:bundleLibs.count];
          for(NSURL *libURL in bundleLibs) {
            NSString *libName = libURL.pathComponents.lastObject;
            if (libName) {
              [bundleLibsNames addObject:libName];
            }
          }
        }

        NSMutableArray *libraries = [[NSMutableArray alloc] init];
        for (NSString* lib in libsList) {
          NSString *libPath = nil;
          if ([bundleLibsNames member:lib]) {
            libPath = [bundleFrameworksFolder stringByAppendingPathComponent:lib];
          } else {
            libPath = [libsFolder stringByAppendingPathComponent:lib];
          }
          [libraries addObject:libPath];
        }

        return libraries;
    }];
}

+ (NSArray<NSURL *> *)filesInDirectory:(NSURL *)directory error:(NSError **)error
{
    NSError *innerError;
    NSArray<NSURL *> *filesInDirectory = [NSFileManager.defaultManager contentsOfDirectoryAtURL:directory includingPropertiesForKeys:@[NSURLIsDirectoryKey] options:0 error:&innerError];
    if (filesInDirectory == nil) {
        *error = [[[FBControlCoreError
                    describeFormat:@"Failed to list files in directory %@", directory]
                   causedBy:innerError]
                  fail:error];
    }
    return filesInDirectory;
}

@end

File: XCTestBootstrap/Utility/FBOToolDynamicLibs.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

enum FBOToolDynamicLibsError: Error {
  case directoryListFailed(path: String, underlying: Error)
  case clangVersionNotFound(path: String)
}

extension FBOToolDynamicLibsError: LocalizedError {
  public var errorDescription: String? {
    switch self {
    case let .directoryListFailed(path, _):
      return "Failed to list files in directory \(path)"
    case let .clangVersionNotFound(path):
      return "No clang version found in \(path)"
    }
  }
}

final class FBOToolDynamicLibs {

  public static func findFullPath(forSanitiserDyldInBundle bundlePath: String) async throws -> [String] {
    let libsList = try await FBOToolOperation.listSanitiserDylibsRequired(byBundle: bundlePath)

    let clangLocation = (FBXcodeConfiguration.developerDirectory as NSString).appendingPathComponent("Toolchains/XcodeDefault.xctoolchain/usr/lib/clang")
    let fileList: [URL]
    do {
      fileList = try FileManager.default.contentsOfDirectory(at: URL(fileURLWithPath: clangLocation), includingPropertiesForKeys: [.isDirectoryKey], options: [])
    } catch {
      throw FBOToolDynamicLibsError.directoryListFailed(path: clangLocation, underlying: error)
    }

    if fileList.isEmpty {
      throw FBOToolDynamicLibsError.clangVersionNotFound(path: clangLocation)
    }

    let libsFolder = NSString.path(withComponents: [fileList[0].path, "lib/darwin/"])

    var bundleFrameworksFolder = (bundlePath as NSString).appendingPathComponent("Frameworks")
    if !FileManager.default.fileExists(atPath: bundleFrameworksFolder) {
      bundleFrameworksFolder = (bundlePath as NSString).appendingPathComponent("Contents/Frameworks")
    }

    var bundleLibsNames: Set<String>?
    if let bundleLibs = try? FileManager.default.contentsOfDirectory(at: URL(fileURLWithPath: bundleFrameworksFolder), includingPropertiesForKeys: [.isDirectoryKey], options: []) {
      var names = Set<String>()
      for libURL in bundleLibs {
        if let libName = libURL.pathComponents.last {
          names.insert(libName)
        }
      }
      bundleLibsNames = names
    }

    var libraries: [String] = []
    for lib in libsList {
      let libPath: String
      if bundleLibsNames?.contains(lib) == true {
        libPath = (bundleFrameworksFolder as NSString).appendingPathComponent(lib)
      } else {
        libPath = (libsFolder as NSString).appendingPathComponent(lib)
      }
      libraries.append(libPath)
    }

    return libraries
  }
}
